File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/LooperThread.java
// SPDX-License-Identifier: GPL-2.0-or-later

package org.dolphinemu.dolphinemu.utils;

import android.os.Looper;

public class LooperThread extends Thread
{
  private Looper mLooper;

  public LooperThread()
  {
    super();
  }

  public LooperThread(String name)
  {
    super(name);
  }

  @Override
  public void run()
  {
    Looper.prepare();

    synchronized (this)
    {
      mLooper = Looper.myLooper();
      notifyAll();
    }

    Looper.loop();
  }

  public Looper getLooper()
  {
    if (!isAlive())
    {
      throw new IllegalStateException();
    }

    synchronized (this)
    {
      while (mLooper == null)
      {
        try
        {
          wait();
        }
        catch (InterruptedException ignored)
        {
        }
      }
    }

    return mLooper;
  }
}

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/LooperThread.kt
// SPDX-License-Identifier: GPL-2.0-or-later

package org.dolphinemu.dolphinemu.utils

import android.os.Looper

class LooperThread @JvmOverloads constructor(name: String? = null) : Thread(name) {
    private val looperLock = Object()
    private var looperInternal: Looper? = null

    override fun run() {
        Looper.prepare()

        synchronized(looperLock) {
            looperInternal = checkNotNull(Looper.myLooper())
            looperLock.notifyAll()
        }

        Looper.loop()
    }

    val looper: Looper
        get() {
            if (!isAlive) {
                throw IllegalStateException("Thread must be started before retrieving the Looper.")
            }

            synchronized(looperLock) {
                while (looperInternal == null) {
                    try {
                        looperLock.wait()
                    } catch (ignored: InterruptedException) {
                    }
                }

                return looperInternal!!
            }
        }
}
